hack/licence-detector/detector/detector.go

```diff
@@ -0,0 +1,145 @@
+package detector
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"path/filepath"
+	"regexp"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/karrick/godirwalk"
+)
+
+var errLicenceNotFound = errors.New("failed to detect licence")
+
+type Dependencies struct {
+	Direct   []LicenceInfo
+	Indirect []LicenceInfo
+}
+
+type LicenceInfo struct {
+	Module
+	LicenceFile string
+	Error       error
+}
+
+type Module struct {
+	Path     string     // module path
+	Version  string     // module version
+	Main     bool       // is this the main module?
+	Time     *time.Time // time version was created
+	Indirect bool       // is this module only an indirect dependency of main module?
+	Dir      string     // directory holding files for this module, if any
+	Replace  *Module    // replace directive
+}
+
+func Detect(data io.Reader, includeIndirect bool) (*Dependencies, error) {
+	dependencies, err := parseDependencies(data, includeIndirect)
+	if err != nil {
+		return nil, err
+	}
+
+	err = detectLicences(dependencies)
+	return dependencies, err
+}
+
+func parseDependencies(data io.Reader, includeIndirect bool) (*Dependencies, error) {
+	deps := &Dependencies{}
+	decoder := json.NewDecoder(data)
+	for {
+		var mod Module
+		if err := decoder.Decode(&mod); err != nil {
+			if err == io.EOF {
+				return deps, nil
+			}
+			return deps, fmt.Errorf("failed to parse dependencies: %w", err)
+		}
+
+		if !mod.Main && mod.Dir != "" {
+			if mod.Indirect {
+				if includeIndirect {
+					deps.Indirect = append(deps.Indirect, LicenceInfo{Module: mod})
+				}
+				continue
+			}
+			deps.Direct = append(deps.Direct, LicenceInfo{Module: mod})
+		}
+	}
+
+	sort.Slice(deps.Direct, func(i, j int) bool {
```

Comment:
I'm not sure when the previous loop breaks and reach this code.

Reply:
Good catch. It's a merge issue that I missed.